src/proofs/weighted_inner_product.rs

```diff
@@ -0,0 +1,915 @@
+                    let bL_minusx = BigInt::mod_mul(&b_L[i], &x_inv_fe.to_big_int(), &order);
+                    BigInt::mod_add(&bRx, &bL_minusx, &order)
+                })
+                .collect::<Vec<BigInt>>();
+            //    b = &mut b_new[..];
+
+            let x2_dL = BigInt::mod_mul(&x_sq_bn, &d_L_fe.to_big_int(), &order);
+            let xinv2_dR = BigInt::mod_mul(&x_sq_inv_bn, &d_R_fe.to_big_int(), &order);
+            let x2_dL_xinv2_dR = BigInt::mod_add(&x2_dL, &xinv2_dR, &order);
+            let alpha_new = BigInt::mod_add(&alpha, &x2_dL_xinv2_dR, &order);
+
+            let x_yinv = BigInt::mod_mul(&x_bn, &powers_yinv[n - 1], &order);
+            let x_yinv_fe = ECScalar::from(&x_yinv);
+            let G_new = (0..n)
+                .map(|i| {
+                    let GLx_inv = &G_L[i] * &x_inv_fe;
+                    let GRx_yinv = &G_R[i] * &x_yinv_fe;
+                    GRx_yinv + GLx_inv
+                })
+                .collect::<Vec<GE>>();
+            //   G = &mut G_new[..];
+
+            let H_new = (0..n)
+                .map(|i| {
+                    let HLx = &H_L[i] * &x;
+                    let HRx_inv = &H_R[i] * &x_inv_fe;
+                    HLx + HRx_inv
+                })
+                .collect::<Vec<GE>>();
+            //    H = &mut H_new[..];
+
+            L_vec.push(L);
+            R_vec.push(R);
+            return WeightedInnerProdArg::prove(
+                &G_new, &H_new, &g, &h, &P, &a_new, &b_new, &alpha_new, L_vec, R_vec,
+            );
+        } else {
+            let r: FE = ECScalar::new_random();
+            let r_bn: BigInt = r.to_big_int();
+            let s: FE = ECScalar::new_random();
+            let s_bn: BigInt = s.to_big_int();
+            let delta: FE = ECScalar::new_random();
+            let delta_bn: BigInt = delta.to_big_int();
+            let eta: FE = ECScalar::new_random();
+            let eta_bn: BigInt = eta.to_big_int();
+
+            // compute A
+            let Gr = &G[0] * &r;
+            let Hs = &H[0] * &s;
+            let a_s = BigInt::mod_mul(&a[0], &s_bn, &order);
+            let a_sy = BigInt::mod_mul(&a_s, &y, &order);
+            let b_r = BigInt::mod_mul(&b[0], &r_bn, &order);
+            let b_ry = BigInt::mod_mul(&b_r, &y, &order);
+            let a_sy_b_ry = BigInt::mod_add(&a_sy, &b_ry, &order);
+            let g_a_sy_b_ry = g * &ECScalar::from(&a_sy_b_ry);
+            let h_delta = h * &delta;
+            let A = Gr + Hs + g_a_sy_b_ry + h_delta;
+
+            // compute B
+            let r_s = BigInt::mod_mul(&r_bn, &s_bn, &order);
+            let r_sy = BigInt::mod_mul(&y, &r_s, &order);
+            let g_r_sy = g * &ECScalar::from(&r_sy);
+            let h_eta = h * &eta;
+            let B = g_r_sy + h_eta;
+
+            // compute challenge e
+            // let lg_n = L_vec.len();
+            let e = HSha256::create_hash_from_ge(&[&g, &h]);
```

Comment:
why not take L, R as well  ?

Reply:
(not critical)

Reply:
The definitions of A and B (as used in paper) are in a form different than L, R. So I have just been consistent with the notation in the paper.

Reply:
Oh I misread your comment. I will include `A` and `B` in generation of the challenge `e`.